README.md
# NAME

ChartJS::Generator - generate HTML chart with Chart.js

# SYNOPSIS

    use ChartJS::Generator;  

    my $c = ChartJS::Generator->new(  
        'Bar', # Bar chart  
        js     => '/path/to/Chart.js',  
        title  => 'Bar chart sample', # HTML title  
        width  => '800', # canvas width  
        height => '600', # canvas height  
    );  

    # create each chart element  
    my $apple = $c->create_element(apple => 'red'); # (name => color)  

    # chart of sales of the yer  
    my @label = ("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12");  
    $c->sort_labels_as_number(1);  
    my @data  = (1, 3, 2, 9, 3, 8, 4, 2, 5, 4, 1, 7);  
    for my $i (0..$#label){  
        $apple->up($label[$i] => $data[$i]);  
    }  

    # render HTML  
    print $c->render;  

# DESCRIPTION

ChartJS::Generator is a simple HTML chart generator with [Chart.js](http://www.chartjs.org/).  
Depend only Chart.js.(It doesn't require any Perl Module)  
Maybe it runs every where.  

# METHOD

### new ( ChartJS::Generator->new($CHART_TYPE, %OPTION))

    $CHART_TYPE    Line, Bar, Radar, PolarArea, Pie, Doughnut  
    %OPTION        js     path to Chart.js for '<script src="">' (default './Chart.js')  
                   title  HTML title  
                   width  canvas width  (default 960)  
                   height canvas height (default 480)  
                   Others are same as Chart.js options. Specify camel case.  

### create_element ($c->create_element($NAME => $COLOR))

    Create a new chart element.  
    $NAME          Name of the chart element. Must be unique.  
    $COLOR         yellow, blue, green, pink, red  

### create_element ($c->create_element($NAME => %COLOR))

    You can also spcify color as hash below.  

    %COLOR         # Line Bar Radar  
                   fill_color   => 'rgba(0,102,255,0.5)',  
                   stroke_color => '#0033ff',  
                   point_color  => '#0000ff',  
                   point_stroke_color => '#fff',  

                   # PolarArea Pie Doughnut  
                   color => 'rgba(255,0,0,0.5)'  

### labels ($c->labels(@LABELS))

    @LABELS        Specify the labels. Any other label in `up` or `down` method ignored.  
                   Labels are sorted in the order of Array.  

### sort_labels_as_number ($c->sort_labels_as_number($BOOL))

    $BOOL          Flag that treat as number when sort labels.  

### reverse_labels ($c->reverse_labels($BOOL))

    $BOOL          Flag that reverse labels or not.  

### up ($elm->up($LABEL => $COUNT) Line, Bar, Radar

    $LABEL         Data point label.  
    $COUNT         Incremental data count of the point. (default 1)  

### up ($elm->up($COUNT)) PolarArea, Pie, Doughnut

    $COUNT         Incremental data count of the point. (default 1)  

### down ($elm->down($LABEL => $COUNT) Line, Bar, Radar

    $LABEL         Data point label.  
    $COUNT         Decremental data count of the point. (default 1)  

### down ($elm->down($COUNT)) PolarArea, Pie, Doughnut

    $COUNT         Decremental data count of the point. (default 1)  

### render ($c->render())

    Render Chart HTML.  

# LICENSE

Copyright (C) Tooru Iwasaki.

This library is free software; you can redistribute it and/or modify  
it under the same terms as Perl itself and under the MIT license as  
Chart.js.  

# AUTHOR

Tooru Iwasaki <rockbone.g@gmail.com>


Changes
Revision history for Perl extension ChartJS-Generator

{{$NEXT}}

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'ChartJS-Generator',
    module_name     => 'ChartJS::Generator',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


sample/bar_chart.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib '../lib';
use ChartJS::Generator;

### Sample data
my @people = qw/toru ken hanako/;
my @term  = qw/1 2 3 4 5 6 7 8 9 10 11 12/;
my %contract_data  = (
    toru   => [ 4, 3, 4, 3, 2, 3, 3, 5, 4, 6, 4, 4 ],
    ken    => [ 1, 3, 3, 3, 2, 3, 7, 7, 2, 3, 2, 1 ],
    hanako => [ 3, 5, 3, 6, 0, 2, 1, 3, 3, 2, 2, 8],
);
my %color = (
    toru      => 'yellow',
    ken       => 'green',
    hanako    => 'red',
);
my $c = ChartJS::Generator->new(
    'Bar',
    js     => 'http://cdn.rockbone.info/js/misc/Chart.js',
    title  => "Corporation Contract 2013",
    width  => 1200,
    height => 800,
);
$c->sort_labels_as_number(1);
for my $p (@people){
    my $elm = $c->create_element($p => $color{$p});
    for my $i (0..$#term){
        $elm->up($term[$i], $contract_data{$p}->[$i]);
    }
}
print $c->render;


sample/radar_chart.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib '../lib';
use ChartJS::Generator;

### Sample data
my @people  = qw/tom tim bob/;
my @category = qw/physical mathematical latent creative productive supernatural/;
my %ability  = (
    tom  => [ 40, 30, 60, 90, 70, 90 ],
    tim  => [ 80, 20, 40, 50, 80, 10 ],
    bob  => [ 60, 90, 70, 80, 20, 40 ],
);
my %color = (
    tom      => 'yellow',
    tim      => 'green',
    bob      => 'pink',
);
my $c = ChartJS::Generator->new(
    'Radar',
    js     => 'http://cdn.rockbone.info/js/misc/Chart.js',
    title  => "Corporation Contract 2013",
    width  => 1200,
    height => 800,
    scaleLineColor => "rgba(1,0,0,.5)"
);
for my $p (@people){
    my $elm = $c->create_element($p => $color{$p});
    for my $i (0..$#category){
        $elm->up($category[$i], $ability{$p}->[$i]);
    }
}
print $c->render;


sample/line_chart_sample.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib '../lib';
use ChartJS::Generator;

### Sample data
my @fluit = qw/apple banana kiwi/;
my @term  = qw/01 02 03 04 05 06 07 08 09 12/;
my %data  = (
    apple  => [ 40,  20,  60, 100,  60,  40,  80, 120,  30,  10, 140,  20],
    banana => [ 60,  30, 120, 100,  70,  90, 120,  40,  50,  50,  70,  90],
    kiwi   => [120,  80,  20,  10,  70, 120,  40,  40,  60, 120, 100,  80],
);
my %color = (
    apple     => 'red',
    banana    => 'yellow',
    kiwi      => 'green',
);
my $c = ChartJS::Generator->new(
    'Line',
    js     => 'http://cdn.rockbone.info/js/misc/Chart.js',
    title  => "Fluit Sales 2013",
    width  => 800,
    height => 500,
);
for my $f (@fluit){
    my $elm = $c->create_element($f => $color{$f});
    for my $i (0..$#term){
        $elm->up($term[$i], $data{$f}->[$i]);
    }
}
print $c->render;


lib/ChartJS/Generator.pm
package ChartJS::Generator;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";

our $CHART_JS      = './Chart.js'; # default <script src=''>
our $CANVAS_WIDTH  = 960;          # default canvas width
our $CANVAS_HEIGHT = 480;          # default canvas height
our @CHART_TYPE    = qw/Line Bar Radar PolarArea Pie Doughnut/;
our @COLOR_NAME    = qw/fill_color stroke_color point_color point_stroke_color color/;
our %COLOR_TABLE   = (
    yellow => ['rgba(255,255,51,0.5)',  'rgba(255,204,51,1)',  'rgba(255,153,51,1)',  '#fff'],
    pink   => ['rgba(255,153,153,0.5)', 'rgba(255,102,153,1)', 'rgba(255,51,153,1)',  '#fff'],
    blue   => ['rgba(102,204,255,0.5)', 'rgba(102,153,255,1)', 'rgba(102,102,255,1)', '#fff'],
    green  => ['rgba(0,255,102,0.5)',   'rgba(0,204,102,1)',   'rgba(0,153,102,1)',   '#fff'],
    red    => ['rgba(204,102,0,0.5)',   'rgba(204,51,0,1)',    'rgba(204,0,0,1)',     '#fff'],
);

# $type  Line|Bar|Radar|PolarArea|Pie|Doughnut
# %opt   js     => /path/to/js
#        title  => HTML title
#        width  => canvas width
#        height => canvas height
#        and others same as Chart.js(specify camel case)
sub new {
    my ($class, $type, %opt) = @_;
    my $self = bless {}, $class;
    ($type) = grep {$type =~ /^$_$/i} @CHART_TYPE;
    die "Unknown chart type [$type]" if !$type;
    $self->type($type);
    $self->js(delete $opt{js});
    $self->title(delete $opt{title});
    $self->width(delete $opt{width});
    $self->height(delete $opt{height});
    $self->options(%opt);
    return $self;
}

sub create_element {
    my ($self, $name) = (shift, shift);
    my $color = @_ == 1        ?  ref $_[0] eq 'HASH' ? $_[0] : $COLOR_TABLE{$_[0]}
              : !@_ || @_ % 2  ? undef
              :                  {@_};
    die "Element's color theme is not specified" if !$color;
    my $element = $self->spawn($name);
    # Refer to COLOR_TABLE
    if (ref $color eq 'HASH'){
        %$color = map {lc($_) => $color->{$_}} keys %$color;
        my @keys = keys %$color;
        if (grep {$self->type eq $_} @CHART_TYPE[0, 1, 2]){
            die "Element's color option is wrong" if @keys != 4 || grep {my $k = $_; !grep {$k eq $_} @COLOR_NAME} @keys;
            $color = [@$color{@COLOR_NAME}];
        }
        else{
            die "Element's color option is wrong" if !exists $color->{color};
            $color = [$color->{color}];
        }
    }
    for (0..3){
        my $setter = $COLOR_NAME[$_];
        $element->$setter($color->[$_]);
    }
    $element->color($color->[0]); # for PolarArea, Pie, Doughnut
    return $element;
}

sub button_as_html {
    my ($self) = @_;
    my $buttons = "";
    for my $elm (@{$self->{elements}}){
        $buttons .= qq(<button style="background-color:@{[ $elm->stroke_color || $elm->color ]};" disabled>@{[ $elm->name ]}</button>);
    }
    return $buttons;
}
sub data_as_json {
    my ($self) = @_;
    my $data;
    if (grep {$self->type eq $_} @CHART_TYPE[0, 1, 2]){ # Line Bar Radar
        $data->{labels} = $self->labels;
        for my $elm (@{$self->{elements}}){
             my $elm_data = {map {snake2camel($_) => $elm->{$_}} qw/fill_color stroke_color point_color point_stroke_color/};
             $elm_data->{data} = [map {$elm->{data}{$_} || 0} $self->labels];
             push @{$data->{datasets}}, $elm_data;
        }
    }
    else{
        for my $elm (@{$self->{elements}}){
            my $elm_data = {};
            $elm_data->{value} = $elm->{data}{value};
            $elm_data->{color} = $elm->color;
            push @$data, $elm_data;
        }
    }
    return jsonize($data);
}
sub options_as_json {jsonize($_[0]->options)}
sub render {
    my ($self) = @_;
    my $html = $self->template;
    for my $key (qw/title js width height type button_as_html data_as_json options_as_json/){
        $html =~ s/\$$key/$self->$key/eg;
    }
    return $html;
}

# Accessor
sub type    {$_[0]->{type}   = $_[1] || $_[0]->{type}}
sub js      {$_[0]->{js}     = $_[1] || $_[0]->{js}     || $CHART_JS}
sub title   {$_[0]->{title}  = $_[1] || $_[0]->{title}  || ""}
sub width   {$_[0]->{width}  = $_[1] || $_[0]->{width}  || $CANVAS_WIDTH}
sub height  {$_[0]->{height} = $_[1] || $_[0]->{height} || $CANVAS_HEIGHT}
sub options {
    my $self = shift;
    $self->{options} = ref $_[0] eq 'HASH' ? $_[0] : {@_} if @_;
    return $self->{options} || {};
}
sub sort_labels_as_number {$_[0]->{sort_labels_as_number} = $_[1] || $_[0]->{sort_labels_as_number}}
sub reverse_labels        {$_[0]->{reverse_labels} = $_[1] || $_[0]->{reverse_labels}}
sub labels {
    my $self = shift;
    $self->{labels} = \@_ if @_;
    my @labels;
    if ($self->{labels}){
        @labels = @{$self->{labels}};
    }
    else{
        my %label;
        for my $elm (@{$self->{elements}}){
            $label{$_} = 1 for $elm->labels;
        }
        @labels = $self->sort_labels_as_number ? sort {$a <=> $b} keys %label : sort keys %label;
    }
    @labels = reverse @labels if $self->reverse_labels;
    return wantarray ? @labels: \@labels;
}
sub element {
    my ($self, $name) = @_;
    die "Element's name is not specified" if !$name;
    my ($element) = grep {$name eq $_->name} @{$self->{elements}};
    return $element;
}
sub spawn {
    my ($self, $name) = @_;
    die "Element [$name] is already exists" if grep {$name eq $_->name} @{$self->{elements}};
    my $element = bless {name => $name, type => $self->type}, (ref $self) . '::Element'; # bless ChartJS::Generator::Element class
    push @{$self->{elements}}, $element;
    return $element;
}

# Utility
sub escape_json {
    $_[0] =~ s/(['"\\])/\\$1/g if $_[0];
    return $_[0];
}

sub quote_str {
    my $thingy = shift;
    return !defined $thingy           ? "null"  # undef
         : ($thingy ^ $thingy) eq '0' ? $thingy # numeric
         : '"' . $thingy . '"';                 # string
}
sub jsonize {
    my $pl_data = shift;
    return ref $pl_data eq 'HASH'  ? hash2js($pl_data) : array2js($pl_data);
}
sub hash2js {
    my $hash = shift;
    die "Not a hash reference" if ref $hash ne 'HASH';
    my $js_str = "{";
    while (my ($key, $val) = each %$hash){
        my $ref = ref $val;
        if (!$ref){
            $js_str .= qq<$key:@{[$val =~ /true|false|null/ ? $val : quote_str(escape_json($val)) ]},>;
        }
        elsif($ref eq 'HASH'){
            $js_str .= qq|$key:@{[hash2js($val)]},|;
        }
        elsif($ref eq 'ARRAY'){
            $js_str .= qq|$key:@{[array2js($val)]},|;
        }
        else{
            die "Object type `$ref` is not supported for method jsonize";
        }
    }
    $js_str =~ s/,?$/}/;
    return $js_str;
}
sub array2js {
    my $array = shift;
    die "Not a array reference" if ref $array ne 'ARRAY';
    my $js_str = "[";
    for my $val (@$array){
        my $ref = ref $val;
        if (!$ref){
            $js_str .= qq<@{[$val =~ /true|false|null/ ? $val : quote_str(escape_json($val)) ]},>;
        }
        elsif($ref eq 'HASH'){
            $js_str .= qq|@{[hash2js($val)]},|;
        }
        elsif($ref eq 'ARRAY'){
            $js_str .= qq|@{[array2js($val)]},|;
        }
        else{
            die "Object type `$ref` is not supported for method jsonize";
        }
    }
    $js_str =~ s/,?$/]/;
    return $js_str;
}

sub snake2camel {
    (my $camel = $_[0] || '') =~ s/(?<!^)_(\w)/uc($1)/eg;
    return $camel;
}

# $xx part will replace by calling render method
sub template {
    <<'TMP';
<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<title>$title</title>
<script src="$js"></script>
</head>
<body>
$title
<div>
 <canvas id="chart" width="$width" height="$height"></canvas>
</div>
<div>
$button_as_html
</div>
<script>
var ctx = document.getElementById("chart").getContext("2d");
var data = $data_as_json;
var option = $options_as_json;
new Chart(ctx).$type(data, option);
</script>
</body>
</html>
TMP
}

## This is the class for each chart element
package ChartJS::Generator::Element;
use strict;
use warnings;

our $VERSION = $ChartJS::Generator::VERSION;

# Accessor
sub name         {$_[0]->{name}}
sub type         {$_[0]->{type}}
sub data         {$_[0]->{data}}
sub labels       {wantarray ? keys %{$_[0]->{data}} : [keys %{$_[0]->{data}}]}
sub color        {$_[0]->{color}        = $_[1] || $_[0]->{color}}
sub fill_color   {$_[0]->{fill_color}   = $_[1] || $_[0]->{fill_color}}
sub stroke_color {$_[0]->{stroke_color} = $_[1] || $_[0]->{stroke_color}}
sub point_color  {$_[0]->{point_color}  = $_[1] || $_[0]->{point_color}}
sub point_stroke_color {$_[0]->{point_stroke_color} = $_[1] || $_[0]->{point_stroke_color}}

## Argument varies by chart type
#  Require label as first argument when chart type is Line, Bar, Radar.
#  In other type ignore label argumnet.
sub up {
    my $self = shift;
    my ($label, $cnt);
    if ($self->type =~ /^Line|Bar|Radar$/){
        $label = shift or die "Incremental target is not specified";
        $cnt = shift || 1;
        die "Argument Incremental count is not numeric" if ($cnt ^ $cnt) ne '0'; # not numeric
    }
    else{
        $label = "value"; # label for PolarArea, Pie, Doughnut
        $cnt = shift || 1;
        die "Chart type `@{[$self->type]}` requires only numeric argument for increment" if ($cnt ^ $cnt) ne '0';
    }
    $self->{data}{$label} += $cnt;
}
sub down {
    my $self = shift;
    my ($label, $cnt);
    if ($self->type =~ /^Line|Bar|Radar$/){
        $label = shift or die "Decremental target is not specified";
        $cnt = shift || 1;
        die "Argument Incremental count is not numeric" if ($cnt ^ $cnt) ne '0'; # not numeric
    }
    else{
        $label = "value"; # label for PolarArea, Pie, Doughnut
        $cnt = shift || 1;
        die "Chart type `@{[$self->type]}` requires only numeric argument for decrement" if ($cnt ^ $cnt) ne '0';
    }
    $self->{data}{$label} -= $cnt;
}

1;
__END__

=encoding utf-8

=head1 NAME

ChartJS::Generator - It's new $module

=head1 SYNOPSIS

    use ChartJS::Generator;

=head1 DESCRIPTION

ChartJS::Generator is ...

=head1 LICENSE

Copyright (C) Tooru Iwasaki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Tooru Iwasaki E<lt>rockbone.g@gmail.comE<gt>

=cut

